Change namespace of attribute classes from `Annotations` to `Attribute`

GromNaN opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no (deprecation)
| Fixed issues | Related https://github.com/doctrine/mongodb-odm/pull/2931

#### Summary

- Move the classes from `Doctrine\ODM\MongoDB\Mapping\Annotations` to `Doctrine\ODM\MongoDB\Mapping\Attribute`. Removing the `@Annotation` so that they can't be used for legacy annotations.
- For the classes in the `Annotations` namespace, they extend those from the `Attribute` namespace. Adding the legacy `@Annotation` for BC.
- The `EncryptQuery` enum is moved to the parent `Doctrine\ODM\MongoDB\Mapping` namespace
- The abstract annotation classes are aliased to the new abstract attribute classes, so that it's not breaking type comparison
- Documentation updated

Related initiative in Symfony: https://github.com/symfony/symfony/issues/51381

GromNaN committed:
Move attribute classes from Mapping\Annotations to Mapping\Attribute namespace

GromNaN committed:
Add deprecated aliases for attribute classes in Annotations namespace

GromNaN commented:
This will force users to migrate without offering anything in return, except for consistency in the namespace with other projects (and not yet ORM). ~We can only do this if ORM also decides to do so.~

GromNaN commented:
In fact, ORM don't have this subnamespace (example: [`Doctrine\ORM\Mapping\Entity`](https://github.com/doctrine/orm/blob/3.5.x/src/Mapping/Entity.php)). So, we should consider whether we want to keep the namespace or move all the attribute classes to `Doctrine\ODM\MongoDB\Mapping`. 
@alcaeus you had the idea that we should use the attribute classes as model classes for all the class metadata. In that case, that would make sense to remove the subnamespace `Attribute`.

malarzm reviewed (commented):
I can't quite grasp it, but there's a "class_alias was problematic in the past" dangling in my head. Saying it out loud in case I'd jog somebody's memory with it. Alernative would be to:

1. Move classes to Attributes (and drop final, leave it as a phpdoc)
2. Classes in Annotations would extend their counterpart, emit a deprecation in ctor and forward everything
3. Remove Annotations/* in 3.0 and bring back `final` in Attribute/*

Also I'd like to see docs updated for 2.16 - folks copying documentation should not be greeted by a deprecation :)

greg0ire commented:
> I can't quite grasp it, but there's a "class_alias was problematic in the past" dangling in my head. Saying it out loud in case I'd jog somebody's memory with it.

@malarzm I wrote [an article about that](https://dev.to/greg0ire/how-to-deprecate-a-type-in-php-48cf), and  @alcaeus wrote [a medium article](https://alcaeus.medium.com/how-to-break-an-entire-ecosystem-by-publishing-a-release-b6aaab2b8aaa) about the aftermath of doing this for `doctrine/persistence` as well.

malarzm commented:
@greg0ire ha! Thanks for the links, I tried to look in our blog and google but prolly I was too specific with "orm namespace change" :)

IonBazan commented on src/Mapping/Annotations/File/ChunkSize.php:
```diff
@@ -4,19 +4,11 @@
+
+trigger_deprecation('doctrine/mongodb-odm', '2.16', 'Namespace "Doctrine\ODM\MongoDB\Mapping\Annotations" is deprecated, use "Doctrine\ODM\MongoDB\Mapping\Attribute" instead.');
+
+class_exists(ChunkSize::class);
```
Because we are only moving `Annotations`, can we opt to a simplified version of deprecation by extending the new classes instead of creating aliases?

```php
/**
 * @deprecated
 */
class ChunkSize extends Doctrine\ODM\MongoDB\Mapping\Attribute\ChunkSize
{
}
```

I know this approach is a BC break when we return such object but I don't think we ever do that. This also works better with IDE autocompletion and composer classmap.

Otherwise, I don't think current approach is aligned with articles linked above either. @malarzm @greg0ire perhaps you have more context.

Something similar was applied to `Route` annotation in Symfony: https://github.com/sensiolabs/SensioFrameworkExtraBundle/blob/2f886f4b31f23c76496901acaedfedb6936ba61f/src/Configuration/Route.php#L14-L24

GromNaN replied:
Using `class_alias` is a lot simpler than removing the final, extending the class and duplicate all the `#[Attribute]` and `/** @Annotation */` annotations.

I've re-added the annotation classes for static analysis, they are never loaded.

The current approches is inspired by the [Twig](https://github.com/twigphp/Twig/blob/2.x/lib/Twig/BaseNodeVisitor.php) and [Symfony Serializer](https://github.com/symfony/symfony/blob/7.4/src/Symfony/Component/Serializer/Annotation/Context.php) migration.

I don't see any issue with the `class_alias` approach. PHPStorm improved its static analysis and actually interprets the class alias as if it was a class declaration (it seems it was not the case few years ago):
<img width="745" height="218" alt="image" src="https://github.com/user-attachments/assets/f87152cf-8f73-42bc-84e1-6d2ca8577faa" />

GromNaN commented on UPGRADE-2.16.md:
````diff
@@ -1,5 +1,27 @@
+
+```diff
+- use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
++ use Doctrine\ODM\MongoDB\Attribute as ODM;
````
```suggestion
+ use Doctrine\ODM\MongoDB\Mapping\Attribute as ODM;
```

GromNaN committed:
Restore annotation classes for static analysis

This classes are never loaded

GromNaN committed:
Make Annotation classes extend the Attribute classes

The Attribute classes cannot be used as Annotation

Move QueryType enum to the Mapping namespace

Keep the tests on the Annotations namespace

GromNaN commented:
I updated the PR, including the docs to use inheritance. That way the new classes in the `Attribute` namespace cannot be used as annotation.

New PR description:
> - Move the classes from `Doctrine\ODM\MongoDB\Mapping\Annotations` to `Doctrine\ODM\MongoDB\Mapping\Attribute`. Removing the `@Annotation` so that they can't be used for legacy annotations.
> - For the classes in the `Annotations` namespace, they extend those from the `Attribute` namespace. Adding the legacy `@Annotation` for BC.
> - The `EncryptQuery` enum is moved to the parent `Doctrine\ODM\MongoDB\Mapping` namespace
> - The abstract annotation classes are aliased to the new abstract attribute classes, so that it's not breaking type comparison
> - Documentation updated

I've kept the tests unchanged, ensuring full BC of existing code.

IonBazan commented on src/Mapping/Attribute/Document.php:
```diff
@@ -0,0 +1,52 @@
+ * @final
+ */
+#[Attribute(Attribute::TARGET_CLASS)]
+class Document extends AbstractDocument implements MappingAttribute
```
Let's move this to `AbstractDocument`, `AbstractField`, `AbstractIndex`:
```suggestion
class Document extends AbstractDocument
```
Less things to remember about in case we add another attribute which extends that class.

GromNaN committed:
Move MappingAttribute to AbstractDocument

alcaeus commented on src/Mapping/Annotations/File/ChunkSize.php:
```diff
@@ -6,17 +6,15 @@
  */
 #[Attribute(Attribute::TARGET_PROPERTY)]
-final class ChunkSize extends AbstractField
+final class ChunkSize extends \Doctrine\ODM\MongoDB\Mapping\Attribute\File\ChunkSize implements Annotation
```
Nit: adding a use statement with alias could make this a little nicer to read:
```php
use Doctrine\ODM\MongoDB\Mapping\Attribute\File\ChunkSize as ChunkSizeAttribute;

final class ChunkSize extends ChunkSizeAttribute implements Annotation
```

Feel free to leave as is if you don't want to go through all annotation classes.

GromNaN replied:
Thanks Copilot, I'm doing this change with minimal efforts.

alcaeus commented on src/Mapping/Attribute/AbstractDocument.php:
```diff
@@ -0,0 +1,14 @@
+}
+
+// @phpstan-ignore class.notFound
+class_alias(AbstractDocument::class, \Doctrine\ODM\MongoDB\Mapping\Annotations\AbstractDocument::class);
```
Noted that this works in conjunction with the `class_exists` call in the `AbstractDocument` annotation file.

GromNaN replied:
Exact, having the `class_alias` in the same file as the class declaration ensure we always have both declared, even if the attribute is used first.
And the `class_exists` only triggers the autoloader.

alcaeus commented on src/Mapping/Annotations/AbstractDocument.php:
```diff
@@ -4,6 +4,14 @@
+// @phpstan-ignore-next-line if.alwaysFalse
+if (false) {
+    /** @deprecated Use \Doctrine\ODM\MongoDB\Mapping\Attribute\AbstractDocument instead */
+    abstract class AbstractDocument extends \Doctrine\ODM\MongoDB\Mapping\Attribute\AbstractDocument
```
The new class no longer implements the `Annotation` class. We can fix this for the static analysis class, but it won't work for the `class_alias` declaration in the attribute.

```suggestion
    abstract class AbstractDocument extends \Doctrine\ODM\MongoDB\Mapping\Attribute\AbstractDocument implements Annotation
```

GromNaN replied:
The abstract class cannot be used as an annotation. Did you find an issue?

alcaeus replied:
This might create a BC break for people with a custom document annotation, but I'd argue that it's very unlikely someone would do that. If you're fine with the potential BC break, so am I :)

GromNaN replied:
I added a note to the backlog and marked the new abstract attribute classes as internal.

alcaeus commented on src/Mapping/Driver/AttributeDriver.php:
```diff
@@ -6,10 +6,8 @@
-use Doctrine\ODM\MongoDB\Mapping\Annotations\AbstractIndex;
-use Doctrine\ODM\MongoDB\Mapping\Annotations\ShardKey;
-use Doctrine\ODM\MongoDB\Mapping\Annotations\TimeSeries;
+use Doctrine\ODM\MongoDB\Mapping\Annotations\Indexes;
```
This import should be removed. It's used in line 112 and 256 where it replaced an `ODM\Indexes` use.

GromNaN replied:
This class is only in the `Annotations` namespace. I did not create the `Attribute` class because it's already deprecated.

alcaeus replied:
Ah, I see. No worries then.

alcaeus commented on src/Mapping/Driver/AttributeDriver.php:
```diff
@@ -22,6 +20,8 @@
 use ReflectionMethod;
 use ReflectionProperty;
 
+use function array_any;
+use function array_find;
```
I like the use of `array_any` and `array_find` in this class 👏

alcaeus commented on src/Mapping/Driver/AttributeReader.php:
```diff
@@ -37,21 +37,21 @@ public function getPropertyAttributes(ReflectionProperty $property): array
-            // Make sure we only get Doctrine Annotations
-            if (! is_subclass_of($attributeName, Annotation::class)) {
+            // Make sure we only get MongoDB ODM attribute classes
+            if (! is_subclass_of($attributeName, MappingAttribute::class)) {
```
Noted that this doesn't change existing behaviour, as all annotations extend the attribute classes.

alcaeus reviewed (changes requested):
Not sure how to proceed with the missing `Annotation` interface on the attribute class aliases. Found one issue where annotation classes were used in the attribute reader.

GromNaN committed:
Add use statement to Attribute class in Annotation classes

GromNaN reviewed (commented):
> Not sure how to proceed with the missing Annotation interface on the attribute class aliases. Found one issue where annotation classes were used in the attribute reader.

Do you have an example, I don't see the issue.

GromNaN committed:
Document minor BC break

GromNaN commented:
Before merging, we need to update the bundle: 
https://github.com/doctrine/DoctrineMongoDBBundle/blob/9c2ccf08a08fcf739d346810adbbd322ed9c0781/src/DependencyInjection/DoctrineMongoDBExtension.php#L535-L552
